chapters/initialization.txt

```diff
@@ -214,6 +214,17 @@ include::{generated}/api/protos/vkEnumerateInstanceVersion.txt[]
     of Vulkan supported by instance-level functionality, encoded as
     described in <<extendingvulkan-coreversions-versionnumbers>>.
 
+An implementation must: return pname:pApiVersion version that is greater or
+equal to 1.1.0, and less than 2.0.0.
+
```

Comment:
Again, we'd prefer to require the implementation to return values corresponding to published specs, but note that applications should be written to be forward compatible with future versions.

Reply:
Again, I kinda disagree. An application written against Vulkan 1.1 must be able to tolerate if the function returns e.g. 1.2.

Reply:
Hm, for now I went with the extremist wording just to test how it feels...

Reply:
So I think I kind of agree with... both of you?

Implementations should be returning exactly the version they support - not "any arbitrary version". We should make this clear here - CTS is written to test functionality based on this returned value.

But to @krOoze's point, I agree applications should be able to handle any future value corresponding to a real spec, though there's nothing fundamentally stopping them going "whoa hey this version is more than I expected".

So I'd suggest changing it into two parts. Something like:

> The value returned in pname:pApiVersion must: correspond to the version of the Vulkan API specification supported by the implementation.
> Minor versions of the Vulkan 1.x specification are backwards compatible, and so applications should: be written to handle any version in the range [1.1.x,2.x.x)

Reply:
I like where you're going with this, but there are some details to work out. Re:

> The value returned in pname:pApiVersion must: correspond...

It's a nit, but a layer (e.g. devsim) can report "1.1" when running on a 1.2 device. That seems like it ought to be allowed; it certainly doesn't make the implementation non-conformant.

Re:

> applications should: be written to handle any version in the range [1.1.x,2.x.x)

Devices are certainly free to restrict themselves to running on 1.2.x or greater. We might say 

>applications targeting 1.x.y should: be written to accept any API version number in the range [1.x.y,2.0.0)

Reply:
@TomOlson Counternit: Not sure layers are restricted to Vulkan spec. I assumed they are generally allowed to do whatever they want, and so have separate API contract with the application that uses them.

If the layer wants to pretend it is conformant 1.1 device running on 1.2. then it must behave as 1.1 device.

@Tobski 
> though there's nothing fundamentally stopping them going "whoa hey this version is more than I expected".

IDK, if an app wants to cripple itself from running on newer versions, who I am to tell it should not. Does not feel noteworthy. Maybe only concern is if someone accidendally writes `if(version != 1.2) abort();` instead of `if(version < 1.2)`. Not sure if this level of hand-holding is necessary. The link to meaning of a version is in the first sentence. And how semver works is generally known.

Reply:
PS: I did not really want to get into what version the application should "handle". I tried to just say the **value** the application gets might be in the whole 1.X range, and I don't care what the app does with the value as long as it acknowledges this fact. I.e. it might crash for 2.0, but it should work gracefully for 1.X (even if that means quiting).

Basically, conformant Vulkan app should **not** have a construct like

```
if (version == VK_VERSION_1_1) doSomething();
else if (version == VK_VERSION_1_0) doSomethingElse();
// But what about all the other versions? This code is probably not valid.
```

But on a second thought I should probably not be telling the app what to do at all. If it is programmed to crash, that is its own business, as long as it does not feed Vulkan API nonsense.

Reply:
@TomOlson 

> It's a nit, but a layer (e.g. devsim) can report "1.1" when running on a 1.2 device. That seems like it ought to be allowed; it certainly doesn't make the implementation non-conformant.

Sure, I meant more "you can't advertise 1.2 if you don't support 1.2 functionality". Layers aren't affected by this language when doing that sort of thing since they are supposed to promise a conformant implementation - it doesn't have to be the same conformant version. A layer advertising 1.1 in this situation just has to make sure that all 1.1 functionality is advertised, and behavior is as expected for a 1.1 implementation.

> applications targeting 1.x.y should: be written to accept any API version number in the range [1.x.y,2.0.0)

That would be nonsense when reading the 1.1 spec. But sure, I take the point - 1.1.x is an example relevant to the current spec only - I had assumed the final language would be more precise.

@krOoze 

> IDK, if an app wants to cripple itself from running on newer versions, who I am to tell it should not.

In this case, the "I" is the Vulkan specification 😄. Should represents a recommendation - applications can feel free to not follow it 🤷‍♀️

So....

...I think we're in agreement on this sentence?

> The value returned in pname:pApiVersion must: correspond to the version of the Vulkan API specification supported by the implementation.

And after thinking about it more, I think I agree the second sentence I suggested is probably redundant and not super useful - so yea maybe just that one?

Reply:
@tobski, I like it, but under the Khronos definition of "support" (which exists outside the specification), an implementation can support multiple versions - so a conformant 1.1 driver supports both the Vulkan 1.1 and 1.0 specifications, and talking about "***the*** version of the Vulkan API specification supported..." is logically incorrect. An implementation that is conformant to the Vulkan 1.2 specification could return 1.2 *or* the version number of any backward-compatible successor version - notionally 1.3, 1.4, et al. 

We could say:
> `must:` correspond to the highest backward-compatible version of the Vulkan API specification supported by the implementation

That statement tells the app writer that they may get a version number higher than 1.2, and the "Extending Vulkan" chapter tells them what version numbers they can confidently expect to be compatible with. The statement doesn't actually constrain the implementer at all, because the only versions of the API the implementation *can* support in the Khronos sense are the one corresponding to pname:pApiVersion and its predecessors.

Reply:
It is odd that we are getting into a prolonged discussion over something that should be a dead simple versioning scheme. I should probably rethink it, so do not waste too much time reviewing the current version.

> An implementation that is conformant to the Vulkan 1.2 specification could return 1.2 or the version number of any backward-compatible successor version - notionally 1.3, 1.4, et al.

I don't follow. An implementation written against the 1.2 Vulkan specification cannot simply return 1.3+.